backend-elasticsearch/src/main/java/org/hibernate/search/v6poc/backend/elasticsearch/search/predicate/impl/BooleanJunctionPredicateBuilderImpl.java

```diff
@@ -85,25 +94,43 @@ private void addMinimumShouldMatchConstraint(int ignoreConstraintCeiling,
 		}
 	}
 
-	private void must(JsonObject query) {
-		MUST.add( getInnerObject(), query );
+	private void must(SearchPredicateContributor<Void, ? super ElasticsearchSearchPredicateCollector> contributor) {
+		if ( mustContributors == null ) {
+			mustContributors = new ArrayList<>();
+		}
+		mustContributors.add( contributor );
 	}
 
-	private void mustNot(JsonObject query) {
-		MUST_NOT.add( getInnerObject(), query );
+	private void mustNot(SearchPredicateContributor<Void, ? super ElasticsearchSearchPredicateCollector> contributor) {
+		if ( mustNotContributors == null ) {
+			mustNotContributors = new ArrayList<>();
+		}
+		mustNotContributors.add( contributor );
 	}
 
-	private void should(JsonObject query) {
-		SHOULD.add( getInnerObject(), query );
+	private void should(SearchPredicateContributor<Void, ? super ElasticsearchSearchPredicateCollector> contributor) {
+		if ( shouldContributors == null ) {
+			shouldContributors = new ArrayList<>();
+		}
+		shouldContributors.add( contributor );
 	}
 
-	private void filter(JsonObject query) {
-		FILTER.add( getInnerObject(), query );
+	private void filter(SearchPredicateContributor<Void, ? super ElasticsearchSearchPredicateCollector> contributor) {
+		if ( filterContributors == null ) {
+			filterContributors = new ArrayList<>();
+		}
+		filterContributors.add( contributor );
 	}
 
 	@Override
 	public void contribute(Void context, ElasticsearchSearchPredicateCollector collector) {
 		JsonObject innerObject = getInnerObject();
```

Comment:
Ok, so this means, at least on the Elasticsearch side, you are re-using the same JsonObject each time `contribute()` is called.

This was not really a problem before, since `contribute()` was only supposed to be called once, but now it could theoretically be called multiple times, right? Which means we could have some issues with clauses being added twice, among others.

What do you think about adding some flag in the superclass to mark the contribute method as called, and throw an exception if called twice? Just to be safe?
We'd need to add a call to `super.contribute(context, collector)` here and in other implementations, of course.

Reply:
Yes, that's what worried me and why I asked you to take a closer look to this part.

Makes sense to add a safeguard. Will do.